pymysql.err.OperationalError: (2013, 'Lost connection to MySQL server during query')

While running the trader, I saw this error multiple times. The situation that leads up to it is as follows, 

1. Socket Manager receives a match for an order in the book, 
2. ThreadHandler checks the match in the Book
3. In the BookManager, Order is found
4. Order filled size is updated
5. If Order filled matches order size, order is marked as filled
6. In the ThreadHandler, Order is again checked to see if it has been marked filled
7. during this check sqlalchemy runs the following query:
```SELECT orders.exchange_id AS orders_exchange_id, orders.pair AS orders_pair, orders.side AS orders_side, orders.price AS orders_price, orders.size AS orders_size, orders.filled AS orders_filled, orders.status AS orders_status, orders.post_only AS orders_post_only, orders.id AS orders_id, orders.created_at AS orders_created_at, orders.book_id AS orders_book_id 
FROM orders 
WHERE orders.id = %(param_1)s]
[parameters: {'param_1': 57768}]
```

8. pymysql engine recieves an IOExeption from MYSQL:  (2013, 'Lost connection to MySQL server during query')

After researching this error, I cant say for cetrain what it is. The query doesnt seem to be a time out, as when I ran it from mysql, it came back without issue. 

One theory is that this query is in the query_cache and upon being updated and imidiately queried again, there is some issue where the query_cache is causing problems due to the update. 

To address the issue, add better error handling and attempt to disconnect and reconnect from the database and try again. Enable logging for mysql in and check trading logs for occurances to cross refrence to mysql. 

```2019-08-12 02:49:50,414 [INFO] trader.socket.reader: < 2019-08-12T08:49:50.870000Z - ETH-USD - trade_id: 50768950 - side: buy size: 0.07140000 price: 211.27000000
2019-08-12 02:49:50,415 [INFO] trader.socket.thread_handler: Checking book for matched trade
2019-08-12 02:49:50,742 [INFO] trader.socket.thread_handler: Matched trade is in book
2019-08-12 02:49:50,773 [ERROR] trader.socket.manager: Socket had a general exception, last message received 0.36115169525146484 seconds ago
Traceback (most recent call last):
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
    cursor, statement, parameters, context
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/cursors.py", line 170, in execute
    result = self._query(query)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/cursors.py", line 328, in _query
    conn.query(q)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 517, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 732, in _read_query_result
    result.read()
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 1075, in read
    first_packet = self.connection._read_packet()
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 657, in _read_packet
    packet_header = self._read_bytes(4)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 707, in _read_bytes
    CR.CR_SERVER_LOST, "Lost connection to MySQL server during query")
pymysql.err.OperationalError: (2013, 'Lost connection to MySQL server during query')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/pi/Projects/Crypto-Liquid-Market-Maker/trader/socket/manager.py", line 121, in listen
    self.reader.new(message)
  File "/home/pi/Projects/Crypto-Liquid-Market-Maker/trader/socket/reader.py", line 31, in new
    return function_name(msg)
  File "/home/pi/Projects/Crypto-Liquid-Market-Maker/trader/socket/reader.py", line 88, in match
    self.check_book(msg)
  File "/home/pi/Projects/Crypto-Liquid-Market-Maker/trader/socket/reader.py", line 108, in check_book
    self.thread_handler.check_book_for_match(msg)
  File "/home/pi/Projects/Crypto-Liquid-Market-Maker/trader/socket/thread_handler.py", line 46, in check_book_for_match
    if order.status == "filled":
  File "/home/pi/Projects/Crypto-Liquid-Market-Maker/trader/exchange/order.py", line 106, in status
    return self._status
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/orm/attributes.py", line 276, in __get__
    return self.impl.get(instance_state(instance), dict_)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/orm/attributes.py", line 677, in get
    value = state._load_expired(state, passive)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/orm/state.py", line 660, in _load_expired
    self.manager.deferred_scalar_loader(self, toload)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 979, in load_scalar_attributes
    only_load_props=attribute_names,
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 208, in load_on_ident
    identity_token=identity_token,
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 282, in load_on_pk_identity
    return q.one()
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 3275, in one
    ret = self.one_or_none()
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 3244, in one_or_none
    ret = list(self)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 3317, in __iter__
    return self._execute_and_instances(context)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 3342, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1248, in _execute_context
    e, statement, parameters, cursor, context
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1466, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 383, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 128, in reraise
    raise value.with_traceback(tb)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
    cursor, statement, parameters, context
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1466, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 383, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 128, in reraise
    raise value.with_traceback(tb)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
    cursor, statement, parameters, context
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/cursors.py", line 170, in execute
    result = self._query(query)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/cursors.py", line 328, in _query
    conn.query(q)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 517, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 732, in _read_query_result
    result.read()
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 1075, in read
    first_packet = self.connection._read_packet()
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 657, in _read_packet
    packet_header = self._read_bytes(4)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 707, in _read_bytes
    CR.CR_SERVER_LOST, "Lost connection to MySQL server during query")
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2013, 'Lost connection to MySQL server during query')
[SQL: SELECT orders.exchange_id AS orders_exchange_id, orders.pair AS orders_pair, orders.side AS orders_side, orders.price AS orders_price, orders.size AS orders_size, orders.filled AS orders_filled, orders.status AS orders_status, orders.post_only AS orders_post_only, orders.id AS orders_id, orders.created_at AS orders_created_at, orders.book_id AS orders_book_id 
FROM orders 
WHERE orders.id = %(param_1)s]
[parameters: {'param_1': 57768}]
(Background on this error at: http://sqlalche.me/e/e3q8)
2019-08-12 02:49:51,141 [ERROR] trader.socket.manager: last message received 0.7087192535400391 seconds ago/nTraceback (most recent call last):
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_c
ontext
    cursor, statement, parameters, context
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 552, in do_execu
te
    cursor.execute(statement, parameters)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/cursors.py", line 170, in execute
    result = self._query(query)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/cursors.py", line 328, in _query
    conn.query(q)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 517, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 732, in _read_query_re
sult
    result.read()
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 1075, in read
    first_packet = self.connection._read_packet()
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 657, in _read_packet
    packet_header = self._read_bytes(4)
  File "/home/pi/.local/share/virtualenvs/Crypto-Liquid-Market-Maker-z1bk84NN/lib/python3.5/site-packages/pymysql/connections.py", line 707, in _read_bytes
    CR.CR_SERVER_LOST, "Lost connection to MySQL server during query")
pymysql.err.OperationalError: (2013, 'Lost connection to MySQL server during query')```

--- a/trader/__main__.py
+++ b/trader/__main__.py
@@ -86,8 +86,8 @@ def construct_trader_module(terms):
   logger.debug("Constructing: \n{}".format(terms))
   BookManagerMaker = book_manager_maker(terms, trading_api=terms.trading_api)
   reader = Reader(BookManagerMaker)
-  socket_manager = SocketManager(reader, BookManagerMaker, product_ids=[terms.pair],
-                                 send_trades=True)
+  socket_manager = SocketManager(reader, terms.trading_api.test,
+                                 product_ids=[terms.pair], send_trades=True)
   return socket_manager
 
 
@@ -128,7 +128,7 @@ def qa():
       thread.intervene(sys.maxsize, None)
       thread.join()
     logger.info("Canceling orders")
-    socket_manager.book_manager.cancel_all_orders()
+    socket_manager.reader.thread_handler.BookManagerMaker().cancel_all_orders()
     logger.info("Done")
 
 

--- a/trader/database/db.py
+++ b/trader/database/db.py
@@ -17,14 +17,14 @@ class DataAccessLayer:
   def connect(self, test=True):
     if not self.conn_string:
       self.set_connection_string(test)
-    self.engine = create_engine(self.conn_string)
+    self.engine = create_engine(self.conn_string, echo_pool=True,
+                                pool_pre_ping=True)
     BaseWrapper.metadata.create_all(self.engine)
     self.Session_Factory = sessionmaker(bind=self.engine)
     self.Session = scoped_session(self.Session_Factory)
 
   def set_connection_string(self, test=True):
     self.conn_string = _db_connection_string(test)
-
 
 dal = DataAccessLayer()
 

--- a/trader/sequence/book_manager.py
+++ b/trader/sequence/book_manager.py
@@ -4,6 +4,8 @@ import logging.config
 from pprint import pformat
 import time
 
+from pymysql import OperationalError
+
 import config
 from trader.exchange.api_wrapper.exchange_api import ExchangeApi
 from trader.exchange.api_wrapper.noop_trader import NoopApi
@@ -41,11 +43,15 @@ class book_manager_maker:
     session = None
     if persist:
       # if they do not exist, this will create tables
-      dal.connect()
-      session = dal.Session()
-      session.add(book)
-      session.commit()
-      kw["book_id"] = book.id
+      try:
+        dal.connect()
+        session = dal.Session()
+        session.add(book)
+        session.commit()
+        kw["book_id"] = book.id
+      except OperationalError as e:
+        logger.error("Initial db connection failed")
+        raise e
     else:
       kw["book"] = book
 
@@ -82,15 +88,20 @@ class BookManager(AbstractBookManager):
     self.session = None
 
   def load_book(self, session=None):
-    if self.book_id is None:
-      logger.error("persist set to {}".format(self.persist))
-      raise RuntimeError("BookManager has no book id this may be because "
-                         "persist is set to false")
-    if session is None:
-      self.session = dal.Session()
-    else:
-      self.session = session
-    self.book = self.session.query(Book).filter(Book.id == self.book_id).one()
+    try:
+      if self.book_id is None:
+        logger.error("persist set to {}".format(self.persist))
+        raise RuntimeError("BookManager has no book id this may be because "
+                           "persist is set to false")
+      if session is None:
+        self.session = dal.Session()
+      else:
+        self.session = session
+      self.book = self.session.query(Book).filter(Book.id == self.book_id).one()
+    except Exception as e:
+      logger.error("Rolling back book load {}".format(e.args))
+      self.session.rollback()
+      raise e
 
   def close_book(self):
     if self.session is not None:
@@ -100,7 +111,12 @@ class BookManager(AbstractBookManager):
 
   def commit(self):
     if self.session is not None:
-      self.session.commit()
+      try:
+        self.session.commit()
+      except Exception as e:
+        logger.error("Rolling back commit: {}".format(e.args))
+        self.session.rollback()
+        raise ConnectionError
 
   def look_for_order(self, match):
     try:
@@ -201,8 +217,8 @@ class BookManager(AbstractBookManager):
         # Another thread is taking priority stopping or pausing
         while not adjust_queue.empty():
           other_max_size, other_thread = adjust_queue.get()
-          logging.warning("Thread with Initial thread being adjusted by priority"
-                          " {}"
+          logging.warning("Thread with Initial thread being adjusted by "
+                          "priority {}"
                           .format(other_max_size))
           if other_max_size >= max_size:
             # Stop this thread work will be reproduced other thread
@@ -224,7 +240,6 @@ class BookManager(AbstractBookManager):
       self.send_order(order)
       time.sleep(.2)
 
-
   def send_order(self, order):
     self.trading_api.send_order(order)
     if order.status == "rejected":
@@ -245,10 +260,13 @@ class BookManager(AbstractBookManager):
     self.commit()
 
   def update_order(self, order, match_size):
+    order_filled = False
     order.filled += match_size
     if order.filled == order.size:
       order.status = "filled"
+      order_filled = True
     self.commit()
+    return order_filled
 
   def cancel_sent_order(self, order):
     self.trading_api.cancel_order(order)
@@ -279,9 +297,11 @@ class BookManager(AbstractBookManager):
                   .format(side, size))
       return
     elif len(matched_orders) > 1:
-      logger.error("Found more than one trade to cancel: {}"
+      logger.warning("Found more than one trade to cancel: {}"
                    .format(matched_orders))
-      return
+      for o in matched_orders:
+        logger.debug("One of orders found to cancel {}".format(o.exchange_id))
+        self.cancel_sent_order(o)
     order_to_cancel = matched_orders[0]
     logging.debug("Found order to cancel {}"
                   .format(order_to_cancel.exchange_id))
@@ -334,19 +354,21 @@ class BookManager(AbstractBookManager):
 
   def post_at_best_post_only(self, order):
     # if we are buying we want minus, if we are selling we want plus
-    details = self.trading_api.get_product_details(self.pair)
+    details = self.trading_api.get_product_details(self.terms.pair)
     plus_or_minus = -1 if order.side == "buy" else 1
     adjust = plus_or_minus * Decimal(details["quote_increment"])
-    ask, bid = self.trading_api.get_first_book(self.pair)
+    ask, bid = self.trading_api.get_first_book(self.terms.pair)
     order.price = (Decimal(bid[0][0]) + adjust if order.side == "buy" else
                    Decimal(ask[0][0]) + adjust)
     self.send_order(order)
 
   def get_sequence_parameter(self, order_description):
-    side, plus_minus = ("buy", -1) if order_description["side"] == "sell" else ("sell", 1)
+    side, plus_minus = ("buy", -1) if order_description["side"] == "sell" else \
+      ("sell", 1)
     count = int(
       1 + (order_description["size"] - self.get_terms_min_size()) /
       self.get_terms_size_change()
     )
-    first_price = order_description["price"] + plus_minus * self.terms.price_change
+    first_price = order_description["price"] + plus_minus * \
+                  self.terms.price_change
     return side, count, first_price

--- a/trader/sequence/noop_book_manager.py
+++ b/trader/sequence/noop_book_manager.py
@@ -58,7 +58,7 @@ class NoopBookManager(AbstractBookManager):
     self.bm.send_order(order)
 
   def update_order(self, order, match_size):
-    self.bm.update_order(order, match_size)
+    return self.bm.update_order(order, match_size)
 
   def cancel_all_orders(self):
     pass

--- a/trader/socket/manager.py
+++ b/trader/socket/manager.py
@@ -1,5 +1,8 @@
 import asyncio
 import base64
+
+from pymysql import OperationalError
+
 import config
 import hashlib
 import hmac
@@ -8,7 +11,6 @@ import logging
 import logging.config
 import time
 import traceback
-from threading import Thread
 import websockets
 
 
@@ -18,7 +20,7 @@ logger = logging.getLogger(__name__)
 
 class SocketManager:
 
-  def __init__(self, reader, BookManagerMaker,
+  def __init__(self, reader, api_test,
                auth=False,
                url="",
                action='subscribe',
@@ -29,19 +31,17 @@ class SocketManager:
 
     self.reader = reader
     self.auth = auth
-    book_manager = BookManagerMaker()
 
     if url == "":
-      if book_manager.is_trading_api_test():
+      if api_test:
         self.url = config.test_socket
       else:
         self.url = config.socket
     else:
       self.url = url
     logging.debug("SocketManager\n\ttest: {}\n\turl: {}\n\tpair: {}".format(
-      book_manager.is_trading_api_test(), self.url, product_ids
+      api_test, self.url, product_ids
     ))
-    self.book_manager = book_manager
     self.channel = channel
     self.product_ids = product_ids
     if self.channel:
@@ -61,6 +61,7 @@ class SocketManager:
     self.api_key = None
     self.api_passphrase = None
     self.api_secret = None
+    self.fast_error_count = 0
 
   def auth_stamp(self):
     self.api_key = config.api_key
@@ -95,10 +96,23 @@ class SocketManager:
         logger.error("Keyboard Interruption, stopping socket")
         raise
 
-      except Exception:
+      except websockets.InvalidStatusCode as e:
+        logger.error("Might be getting rate limited")
+        if self.last_time_watch() < 3:
+          self.fast_error_count += 1
+        if self.fast_error_count > 5:
+          raise e
+        time.sleep(10)
+
+      except Exception as e:
         logger.error("last message received {} seconds ago/n{}"
                      .format(self.last_time_watch(),
                              traceback.format_exc()))
+        if self.last_time_watch() < 3:
+          self.fast_error_count += 1
+        if self.fast_error_count > 5:
+          logger.error("Too many repeated errors")
+          raise e
         logger.warning("Restarting Web Socket")
 
   async def connect(self):
@@ -119,16 +133,22 @@ class SocketManager:
         message = json.loads(received)
         logger.debug("< {}".format(message))
         self.reader.new(message)
+        self.fast_error_count = 0
         self.last_time = time.time()
 
       except asyncio.TimeoutError:
         await self.ping_socket()
 
       except websockets.exceptions.ConnectionClosed:
-        logging.warn("Connection Closed Exception after {} seconds"
+        logging.warning("Connection Closed Exception after {} seconds"
                      .format(round(self.last_time_watch(), 2)))
         raise websockets.exceptions.ConnectionClosed(
           1006, "1 minute without messages")
+
+      except OperationalError:
+        # initial error will be raised, repeated reconnections will also
+        # be raised, logging will occur at specified level
+        raise
 
       except KeyboardInterrupt:
         raise

--- a/trader/socket/reader.py
+++ b/trader/socket/reader.py
@@ -75,11 +75,12 @@ class Reader:
       }
     :return: none
     """
-    logger.info("< {0} - {1} - trade_id: {2} - "
-                "side: {3} size: {4} price: {5}".format(
+    logger.info("< {} - {} - trade_id: {} - maker_order_id: {}"
+                "side: {} size: {} price: {}".format(
                   msg["time"],
                   msg["product_id"],
                   msg["trade_id"],
+                  msg["maker_order_id"],
                   msg["side"],
                   msg["size"],
                   msg["price"]

--- a/trader/socket/thread_handler.py
+++ b/trader/socket/thread_handler.py
@@ -3,7 +3,10 @@ import queue
 import threading
 from decimal import Decimal
 
+from pymysql import OperationalError
+
 import config
+from trader.database.db import dal
 
 logging.config.dictConfig(config.log_config)
 logger = logging.getLogger(__name__)
@@ -22,6 +25,7 @@ class ThreadHandler:
     self.initial_done = False
     self._sell_threads = []
     self._buy_threads = []
+    self.session_errors = 0
 
   def get_threads(self):
     return self._sell_threads + self._buy_threads + [self.initial_thread]
@@ -29,36 +33,51 @@ class ThreadHandler:
   def check_book_for_match(self, match):
     """Check if matched trade is in book and process trade if it is
     """
-    logger.info("Checking book for matched trade")
-    book_manager = self.BookManagerMaker()
-    book_manager.load_book()
-    order = book_manager.look_for_order(match)
+    try:
+      logger.info("Checking book for matched trade")
+      book_manager = self.BookManagerMaker()
+      book_manager.load_book()
+      order = book_manager.look_for_order(match)
 
-    if order is None:
-      logger.info("Matched trade not in book")
-      return
+      if order is None:
+        logger.info("Matched trade not in book")
+        return
 
-    logger.info("Matched trade is in book")
-    match_size = Decimal(match["size"])
+      logger.info("Matched trade is in book")
+      match_size = Decimal(match["size"])
 
-    book_manager.update_order(order, match_size)
+      is_order_filled = book_manager.update_order(order, match_size)
 
-    if order.status == "filled":
-      logger.info("Match filled order")
-      # order was loaded in current thread, passing it as a dictionary
-      # to avoid session errors caused by passing it directly
-      order_description = {
-        "side": order.side,
-        "size": order.size,
-        "price": order.price
-      }
-      self.handle_full_match(order_description)
+      if is_order_filled:
+        logger.info("Match filled order")
+        # order was loaded in current thread, passing it as a dictionary
+        # to avoid session errors caused by passing it directly
+        order_description = {
+          "side": order.side,
+          "size": order.size,
+          "price": order.price
+        }
+        self.handle_full_match(order_description)
+      else:
+        logger.info("Match filled {} of order, {} is remaining."
+                    .format(Decimal(match_size), order.size - order.filled))
 
-    else:
-      logger.info("Match filled {} of order, {} is remaining."
-                  .format(Decimal(match_size), order.size - order.filled))
-
-    book_manager.close_book()
+      book_manager.close_book()
+      self.session_errors = 0
+    except ConnectionError as e:
+      dal.disconnect()
+      dal.connect()
+      self.session_errors += 1
+      if self.session_errors < 3:
+        logger.warning(
+          "Attempting to reconnect for database exception code: {} message: {}"
+          "".format(
+            e.args[0], e.args[1]
+          )
+        )
+        self.check_book_for_match(match)
+      else:
+        raise e
 
   def handle_full_match(self, order_desc):
     if order_desc["side"] == "buy":
@@ -118,6 +137,7 @@ class TradingThread(threading.Thread):
     self.book_manager = book_manager
     self.adjust_queue = queue.Queue()
     self.started = False
+    self.session_errors = 0
 
   def start(self):
     self.started = True
@@ -127,13 +147,26 @@ class TradingThread(threading.Thread):
     return not self.started
 
   def run(self):
-    self.book_manager.load_book()
-    if self.filled_order_desc is None:
-      self.book_manager.send_orders(adjust_queue=self.adjust_queue)
-    else:
-      self.book_manager.send_trade_sequence(self.filled_order_desc,
-                                            adjust_queue=self.adjust_queue)
-    self.book_manager.close_book()
+    try:
+      self.book_manager.load_book()
+      if self.filled_order_desc is None:
+        self.book_manager.send_orders(adjust_queue=self.adjust_queue)
+      else:
+        self.book_manager.send_trade_sequence(self.filled_order_desc,
+                                              adjust_queue=self.adjust_queue)
+      self.book_manager.close_book()
+    except ConnectionError as e:
+
+      self.session_errors += 1
+      if self.session_errors < 3:
+        logger.warning(
+          "Attempting to reconnect for database exception code: {}"
+          "".format(e.args)
+        )
+        dal.connect()
+        self.run()
+      else:
+        raise e
 
   def intervene(self, size, other_thread):
     """

--- a/trader/test/sequence/unit/test_book_manager.py
+++ b/trader/test/sequence/unit/test_book_manager.py
@@ -189,8 +189,7 @@ class TestBookManager(unittest.TestCase):
     match = create_match(order)
     self.assertTrue(self.book_manager.look_for_order(match), order)
 
-    self.book_manager.update_order(order, match["size"])
-    self.assertTrue(order.status == "filled")
+    self.assertTrue(self.book_manager.update_order(order, match["size"]))
     order_desc = {
       "side": order.side,
       "size": order.size,

--- a/trader/test/socket/unit/test_thread_handler.py
+++ b/trader/test/socket/unit/test_thread_handler.py
@@ -179,7 +179,7 @@ class TestThreadHandler(unittest.TestCase):
         (price is None or o.price == Decimal(price))
       )
     except StopIteration:
-      if time.time() - start < 10:
+      if time.time() - start < 1:
         time.sleep(.005)
         return self.create_match_from_size(size)
       else:

